markdown/inlinepatterns.py

```diff
@@ -431,6 +439,154 @@ def get_stash(m):
+
+        `last` (`etree.Element`):
+            Last appended child to parent. Can also be None if parent has no children.
+
+        `idx` (`int`):
+            Current pattern index that was used to evaluate the parent.
+
+        """
+
+        offset = 0
+        pos = 0
+
+        length = len(data)
+        while pos < length:
+            # Find the start of potential emphasis or strong tokens
+            if self.compiled_re.match(data, pos):
+                matched = False
+                # See if the we can match an emphasis/strong pattern
+                for index, pattern in enumerate(self.PATTERNS):
+                    # Only evaluate patterns that are after what was used on the parent
+                    if index <= idx:
+                        continue
+                    m = pattern[0].match(data, pos)
+                    if m:
+                        # Append child nodes to parent
+                        # Text nodes should be appended to the last
+                        # child if present, and if not, it should
+                        # be added as the parent's text node.
+                        text = data[offset:m.start(0)]
+                        if text:
+                            if last is not None:
+                                last.tail = text
+                            else:
+                                parent.text = text
+                        el = self.build_element(m, pattern[1], pattern[2], index)
+                        parent.append(el)
+                        last = el
+                        # Move our position past the matched hunk
+                        offset = pos = m.end(0)
+                        matched = True
+                if not matched:
+                    # We matched nothing, move on to the next character
+                    pos += 1
+            else:
+                # Increment position as no potential emphasis start was found.
+                pos += 1
+
+        # Append any leftover text as a text node.
+        text = data[offset:]
+        if text:
+            if last is not None:
+                last.tail = text
+            else:
+                parent.text = text
+
+    def build_element(self, m, builder, tags, index):
+        """Element builder."""
+
+        if builder == 'double2':
+            return self.build_double2(m, tags, index)
+        elif builder == 'double':
+            return self.build_double(m, tags, index)
+        else:
+            return self.build_single(m, tags, index)
+
+    def handleMatch(self, m, data):
+        """Parse patterns."""
+
+        el = None
+        start = None
+        end = None
+
+        for index, pattern in enumerate(self.PATTERNS):
+            m1 = pattern[0].match(data, m.start(0))
+            if m1:
+                start = m1.start(0)
+                end = m1.end(0)
+                el = self.build_element(m1, pattern[1], pattern[2], index)
```

Comment:
It took me a minute to realize what `pattern[0]`, `pattern[1]`, and `pattern[2]` were referring to. I wonder if a named tuple would make sense here. That way we would have `pattern.re` (or some better name; `pattern.pattern` seems weird), `pattern.builder`, and `pattern.tags`. That would also clear up what all those strings mean where `PATTERNS` is defined.

In all other respects this looks good, pending complete test coverage.

I hesitated for a minute as we don't get the granularity to override specific parts with extensions with this implementation. But then I remembered that the various parts were all reliant on each other anyway. Altering an earlier pattern often required also altering a later pattern. This actually makes the fact that they are tied together more clear. And it is easy enough for an extension to subclass the Processor class and override the relevant pieces. :+1:

Reply:
> I wonder if a named tuple would make sense here.

I was actually considering doing a named tuple, so I fully agree.

Reply:
> I hesitated for a minute as we don't get the granularity to override specific parts with extensions with this implementation.

I definitely understand this.  This is why I originally brought this discussion up in the related issue.  But ultimately, I felt the best way to get the discussion going was to just get some code up 🙂.

I'm okay if we want to just insert the new pattern between other patterns via the old way.  We lose the pattern steps by 10s, but you retain similar layout. I'm open to either way if we change our mind before merging, but at least now it is clear what the alternative would look like.

Reply:
If I had never seen the old way, I would find this less method surprising that what we have now. We will just need to document the change in the release notes as it could be backward incompatible for some extensions. Unfortunately, there is probably not a way to gracefully deprecate the old patterns.

Although, if `markdown.util.Registry.deregister` defaulted to `strict=False`, that might not be an issue. Of course, it defaults to `strict=true` right now. Not sure I want to change the default though.